src/server/pfs/server/driver_file.go

```diff
@@ -52,64 +52,50 @@ func (d *driver) oneOffModifyFile(ctx context.Context, repo, branch string, cb f
-			if retErr == nil {
-				retErr = d.finishCommit(txnCtx, commit, "")
-			}
-		}()
-		return d.withCommitWriter(txnCtx.ClientContext, commit, cb)
+		if err := d.withCommitUnorderedWriter(txnCtx.Client, commit, cb); err != nil {
+			return err
+		}
+		return d.finishCommit(txnCtx, commit, "")
 	})
 }
 
 // withCommitWriter calls cb with an unordered writer. All data written to cb is added to the commit, or an error is returned.
-func (d *driver) withCommitWriter(ctx context.Context, commit *pfs.Commit, cb func(*fileset.UnorderedWriter) error) (retErr error) {
-	n := d.getSubFileset()
-	subFileSetStr := fileset.SubFileSetStr(n)
-	subFileSetPath := path.Join(commit.Repo.Name, commit.ID, subFileSetStr)
-	return d.storage.WithRenewer(ctx, defaultTTL, func(ctx context.Context, renewer *renew.StringSet) error {
-		id, err := d.withTmpUnorderedWriter(ctx, renewer, false, cb)
+func (d *driver) withCommitUnorderedWriter(pachClient *client.APIClient, commit *pfs.Commit, cb func(*fileset.UnorderedWriter) error) (retErr error) {
+	return d.storage.WithRenewer(pachClient.Ctx(), defaultTTL, func(ctx context.Context, renewer *renew.StringSet) error {
+		id, err := d.withUnorderedWriter(ctx, renewer, false, cb)
 		if err != nil {
 			return err
 		}
-		tmpPath := path.Join(tmpRepo, id)
-		return d.storage.Copy(ctx, tmpPath, subFileSetPath, 0)
+		return d.commitStore.AddFileset(ctx, commit, *id)
 	})
 }
 
-func (d *driver) getSubFileset() int64 {
-	// TODO subFileSet will need to be incremented through postgres or etcd.
-	nonce := atomic.AddUint64(&d.nonce, 1)
-	millis := time.Now().UnixNano() / 1e6 // nano -> milli
-	return millis + int64(nonce%1e6)
-}
-
-func (d *driver) withTmpUnorderedWriter(ctx context.Context, renewer *renew.StringSet, compact bool, cb func(*fileset.UnorderedWriter) error) (string, error) {
-	id := uuid.NewWithoutDashes()
-	inputPath := path.Join(tmpRepo, id)
+func (d *driver) withUnorderedWriter(ctx context.Context, renewer *renew.StringSet, compact bool, cb func(*fileset.UnorderedWriter) error) (*fileset.ID, error) {
 	opts := []fileset.UnorderedWriterOption{fileset.WithRenewal(defaultTTL, renewer)}
-	defaultTag := fileset.SubFileSetStr(d.getSubFileset())
-	uw, err := d.storage.NewUnorderedWriter(ctx, inputPath, defaultTag, opts...)
+	uw, err := d.storage.NewUnorderedWriter(ctx, d.getDefaultTag(), opts...)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	if err := cb(uw); err != nil {
-		return "", err
+		return nil, err
 	}
-	if err := uw.Close(); err != nil {
-		return "", err
+	id, err := uw.Close()
+	if err != nil {
+		return nil, err
 	}
-	if compact {
-		outputPath := path.Join(tmpRepo, id, fileset.Compacted)
-		_, err := d.storage.Compact(ctx, outputPath, []string{inputPath}, defaultTTL)
-		if err != nil {
-			return "", err
-		}
-		renewer.Add(outputPath)
+	if !compact {
+		renewer.Add(*id)
+		return id, nil
 	}
-	return id, nil
+	compactedID, err := d.storage.Compact(ctx, []fileset.ID{*id}, defaultTTL)
```

Comment:
I think we will be able to remove this if we move to a compaction at read time model, so we would just always return the composite fileset. That doesn't really change anything in this PR, I just wanted to make a note.

Reply:
Yeah I agree. I think we need to keep this for the time being to get the reasonable behavior we talked about for temporary filesets and commits.